src/charm.py

```diff
@@ -75,6 +121,89 @@ def _change_primary(self) -> None:
         except SwitchoverFailedError as e:
             logger.error(f"switchover failed with reason: {e}")
 
+    def _on_peer_relation_changed(self, event: RelationChangedEvent):
+        """Reconfigure cluster members when something changes."""
+        # Prevents the cluster to be reconfigured before it's bootstrapped in the leader.
+        if "cluster_initialised" not in self._peers.data[self.app]:
+            event.defer()
+            return
+
+        # If the unit is the leader, it can reconfigure the cluster.
+        if self.unit.is_leader():
+            self._add_members(event)
+
+        # Don't update this member before it's part of the members list.
+        if self._unit_ip not in self.members_ips:
+            return
+
+        # Update the list of the cluster members in the replicas to make them know each other.
+        try:
+            # Update the members of the cluster in the Patroni configuration on this unit.
+            self._patroni.update_cluster_members()
+        except RetryError:
+            self.unit.status = BlockedStatus("failed to update cluster members on member")
+            return
+
+        # Start can be called here multiple times as it's idempotent.
+        # At this moment, it starts Patroni at the first time the data is received
+        # in the relation.
+        self._patroni.start_patroni()
+
+        # Assert the member is up and running before marking the unit as active.
+        if not self._patroni.member_started:
+            self.unit.status = WaitingStatus("awaiting for member to start")
+            event.defer()
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _add_members(self, event):
+        """Add new cluster members.
+
+        This method is responsible for adding new members to the cluster
+        when new units are added to the application. This event is deferred if
+        one of the current units is copying data from the primary, to avoid
+        multiple units copying data at the same time, which can cause slow
+        transfer rates in these processes and overload the primary instance.
+        """
+        try:
+            # Compare set of Patroni cluster members and Juju hosts
+            # to avoid the unnecessary reconfiguration.
+            if self._patroni.cluster_members == self._hosts:
+                return
+
+            logger.info("Reconfiguring cluster")
+            self.unit.status = MaintenanceStatus("reconfiguring cluster")
+            self.unit.status = MaintenanceStatus("reconfiguring cluster")
```

Comment:
Duplicate line

Reply:
Good catch! Thanks. I fixed it in https://github.com/canonical/postgresql-operator/pull/11/commits/73adbd589e3c044a3b19bd846df129db14c68e4c.